### DISCUSSION TITLE
Abstracting out data?

### DISCUSSION THREAD

**Discussion by klevermonicker** [SilentVoid13/Templater#725] [Help] [answered]
I'm trying to make my collection of templates a little more generic and less about me so I can actually share them around without compromising a lot of what my notes are about. I've got a simple file that I have a user javavscript function reading....I can't however figure out how to get this script to either
A) Return the sorted array's to me for a suggester
or 
B) Fire up a suggester with the data prepopulated.

What ends up happening is this error:
VM313:16 Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'system')

What am I doing wrong?

Here's the code itself:
```
async function people(tp) {
var nameParse = '';
var fullName = [];
var firstName = [];
var fs = require('fs');
fs.readFile('../data/people.txt', 'utf8', function(err, data){
    if (typeof data === 'string') { 
    nameParse = data.split(",");
    }
    nameParse = Array.from(nameParse).sort();
    nameParse.forEach(function(item) {
        firstName.push(item.trim().split(" ")[0]);
        fullName.push(item.trim())
    });
})
    var name = await tp.system.suggester(items = firstName, text_items = fullName);
}

module.exports = people

```

**Comment by AB1908** [comment, marked as answer]
`tp` is likely not being passed in properly. Can you log and inspect it?

**Reply by klevermonicker** [reply]
Being honest, I'm pretty new to JS and do things like this...so I'm hoping this is what you meant?

```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'system')
    at Object.people (eval at <anonymous> (plugin:templater-obsidian:1:1), <anonymous>:20:21)
    at eval (eval at compile (plugin:templater-obsidian:1:1), <anonymous>:6:9)
    at eval (plugin:templater-obsidian:3612:54)
    at new Promise (<anonymous>)
    at render (plugin:templater-obsidian:3610:16)
    at renderAsync (plugin:templater-obsidian:3626:10)
    at Parser.eval (plugin:templater-obsidian:3635:23)
    at Generator.next (<anonymous>)
    at eval (plugin:templater-obsidian:63:61)
    at new Promise (<anonymous>)
```

When I click into the eval it shows me that one line with red squigglies. Am I calling the suggester wrong or something?

**Reply by AB1908** [reply]
Can you try following this guide? https://shbgm.ca/obsidian/docs/how-to-use-templater-js-scripts

**Reply by klevermonicker** [reply]
The notice script from there works just fine for me. That's literally where I started working on this.

**Reply by klevermonicker** [reply]
Ok, I understand what you meant now. I got it figured out sort of. Looking at variable scoping most likely now. I wasn't calling this `<%* tp.user.people(tp) %>` I was instead calling `<%* tp.user.people() %>`. Just need to understand why it feels my variables are empty and I should be golden! Thank you!

**Reply by AB1908** [reply]
Haha glad to see you got it working.
